File: chat/web/src/search.test.ts
import { describe, it, expect } from "vitest";
import {
  findMatches,
  fold,
  locate,
  recentChats,
  searchChats,
  snippet,
} from "./search";
import type { Chat, Entry } from "./types";

const entry = (over: Partial<Entry>): Entry => ({
  id: "e",
  role: "assistant",
  text: "",
  detail: "",
  createdAt: 0,
  isStreaming: false,
  delivery: "",
  ...over,
});
const chat = (
  id: string,
  title: string,
  entries: Entry[],
  archived = false,
): Chat => ({
  id,
  title,
  sandboxID: "s",
  repository: "",
  status: "idle",
  archived,
  conversation: { entries },
  approvals: [],
});

describe("folding", () => {
  it("ignores case, accents and the kind of whitespace, keeping offsets", () => {
    expect(fold("Café\tAu Lait\n")).toBe("cafe au lait ");
    expect(fold("Ärger")).toBe("arger");
    for (const s of ["İstanbul", "straße", "𝐀 bold", "e\u0301", "日本語"])
      expect(fold(s).length).toBe(s.length);
    expect(fold("e\u0301")).toBe("e\u0301".toLowerCase());
    expect(fold("𝐀")).toBe("𝐀");
  });
});

describe("findMatches", () => {
  it("finds every literal, case-insensitive, non-overlapping match", () => {
    expect(findMatches("Aaa aA", "aa")).toEqual([
      { start: 0, end: 2 },
      { start: 4, end: 6 },
    ]);
    expect(findMatches("résumé resume", "RESUME")).toEqual([
      { start: 0, end: 6 },
      { start: 7, end: 13 },
    ]);
    expect(findMatches("a\nb", "a b")).toEqual([{ start: 0, end: 3 }]);
    expect(findMatches("x y", " x ")).toEqual([{ start: 0, end: 1 }]);
  });
  it("matches nothing for a blank query and honours the limit", () => {
    expect(findMatches("anything", "")).toEqual([]);
    expect(findMatches("anything", "  \n")).toEqual([]);
    expect(findMatches("", "a")).toEqual([]);
    expect(findMatches("a a a", "a", 2)).toHaveLength(2);
  });
});

describe("locate", () => {
  it("maps an offset into the concatenation of text nodes", () => {
    const lengths = [3, 0, 4];
    expect(locate(lengths, 0)).toEqual({ index: 0, offset: 0 });
    expect(locate(lengths, 2)).toEqual({ index: 0, offset: 2 });
    // A boundary starts in the next node and ends in the previous one.
    expect(locate(lengths, 3)).toEqual({ index: 2, offset: 0 });
    expect(locate(lengths, 3, true)).toEqual({ index: 0, offset: 3 });
    expect(locate(lengths, 7, true)).toEqual({ index: 2, offset: 4 });
    expect(locate([5], 5, true)).toEqual({ index: 0, offset: 5 });
  });
});

describe("snippet", () => {
  it("cuts around the match at word boundaries and collapses whitespace", () => {
    const text =
      "The quick brown fox\njumps over the lazy dog and keeps on running far away";
    const [match] = findMatches(text, "lazy");
    expect(snippet(text, match, 12)).toEqual({
      before: "…over the ",
      match: "lazy",
      after: " dog and…",
    });
    expect(snippet("lazy dog", { start: 0, end: 4 })).toEqual({
      before: "",
      match: "lazy",
      after: " dog",
    });
    expect(snippet("a\n\nlazy", { start: 3, end: 7 }).before).toBe("a ");
  });
});

describe("searchChats", () => {
  const old = chat("old", "Fix the build", [
    entry({ id: "o1", role: "user", text: "Why does the build fail?" }),
    entry({
      id: "o2",
      role: "activity",
      text: "bash",
      detail: "make: *** No rule to make target build",
      createdAt: 5,
    }),
  ]);
  const recent = chat("new", "Deploy notes", [
    entry({ id: "n1", role: "user", text: "Write the deploy notes" }),
    entry({ id: "n2", text: "Notes: the BUILD is green", createdAt: 10 }),
  ]);
  const gone = chat(
    "gone",
    "Build archive",
    [entry({ id: "g1", text: "archived build", createdAt: 20 })],
    true,
  );
  it("orders chats by recency with archived ones last", () => {
    expect(recentChats([old, gone, recent]).map((c) => c.id)).toEqual([
      "new",
      "old",
      "gone",
    ]);
  });
  it("lists title hits first, then entries newest first, one per entry", () => {
    const { hits, more } = searchChats([old, gone, recent], "build");
    expect(more).toBe(0);
    expect(
      hits.map((h) =>
        h.kind === "chat" ? `chat:${h.chat.id}` : `${h.entry.id}:${h.field}`,
      ),
    ).toEqual([
      "chat:old",
      "chat:gone",
      "n2:text",
      "o2:detail",
      "o1:text",
      "g1:text",
    ]);
    const title = hits[0];
    expect(title.kind === "chat" && title.match).toEqual({ start: 8, end: 13 });
  });
  it("matches nothing for a blank query and counts hits past the limit", () => {
    expect(searchChats([old, recent], " ")).toEqual({ hits: [], more: 0 });
    const { hits, more } = searchChats([old, gone, recent], "build", 2);
    expect(hits).toHaveLength(2);
    expect(more).toBe(4);
  });
  it("searches a tool step's output only when its label has no match", () => {
    const { hits } = searchChats([old], "make");
    expect(hits.map((h) => h.kind === "entry" && h.field)).toEqual(["detail"]);
    expect(searchChats([old], "bash").hits[0]).toMatchObject({
      field: "text",
    });
  });
});

File: chat/internal/chats/search_test.go
package chats

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"warden/chat/internal/conversation"
)

// The search reaches every chat's title and transcript, archived chats
// and a subagent's nested entries and the person's own commands
// included, in the palette's order (search.ts): title hits first, then
// entries newest chat and newest entry first; case and the kind of
// whitespace do not matter; the limit caps the hits and counts the rest.
func TestSearchAcrossChats(t *testing.T) {
	e, _, _ := setup(t)
	entry := func(id, role, text, detail string, at float64) conversation.Entry {
		return conversation.Entry{ID: id, Role: role, Text: text, Detail: detail, CreatedAt: at}
	}
	err := e.Store.update(func(st *State) error {
		old := &Chat{ID: "old", Title: "Fix the build", Provider: "claude", Status: "idle"}
		old.Conversation.Entries = []conversation.Entry{
			entry("o1", "user", "Why does the\tBUILD fail?", "", 1),
			entry("o2", "activity", "make", "make: *** No rule to make target build", 5),
		}
		recent := &Chat{ID: "new", Title: "Deploy notes", Provider: "claude", Status: "idle"}
		recent.Conversation.Entries = []conversation.Entry{
			entry("n1", "user", "Write the deploy notes", "", 8),
			entry("n2", "assistant", "Notes: the Build is green", "", 10),
			entry("agent", "activity", "Agent: look (Explore)", "It is in lex.go.", 11),
			{ID: "child", Role: "activity", Text: "Grep \"tokenizer\" in .", Detail: "lex.go:12", ParentID: "agent", CreatedAt: 12, Tool: &conversation.Tool{Kind: "search", Status: "completed"}},
			{ID: "mine", Role: "activity", Text: "git status", Detail: "?? scratch.txt", Sender: &conversation.Actor{PrincipalID: "owner"}, CreatedAt: 13, Tool: &conversation.Tool{Kind: "command", Status: "completed"}},
		}
		gone := &Chat{ID: "gone", Title: "Build archive", Provider: "codex", Status: "idle", Archived: true}
		gone.Conversation.Entries = []conversation.Entry{entry("g1", "assistant", "archived build", "", 20)}
		st.Chats = append(st.Chats, old, recent, gone)
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	keys := func(r SearchResult) []string {
		var out []string
		for _, h := range r.Hits {
			if h.Field == "title" {
				out = append(out, "chat:"+h.ChatID)
			} else {
				out = append(out, h.EntryID+":"+h.Field)
			}
		}
		return out
	}
	r := e.Search("build", 0)
	if got := strings.Join(keys(r), " "); got != "chat:old chat:gone n2:text o2:detail o1:text g1:text" || r.More != 0 {
		t.Fatalf("order: %s (more %d)", got, r.More)
	}
	if h := r.Hits[0]; h.Snippet.Match != "build" || h.Snippet.Before != "Fix the " || h.Title != "Fix the build" {
		t.Fatalf("title hit: %+v", h)
	}
	if h := r.Hits[2]; h.Role != "assistant" || h.CreatedAt != 10 || h.Snippet.Match != "Build" || h.Snippet.After != " is green" || h.Snippet.Before != "Notes: the " {
		t.Fatalf("entry hit: %+v", h)
	}
	if h := r.Hits[4]; h.Snippet.Match != "BUILD" || h.Snippet.Before != "Why does the " {
		t.Fatalf("whitespace folded: %+v", h)
	}
	if h := r.Hits[5]; !h.Archived || h.Provider != "codex" {
		t.Fatalf("archived hit: %+v", h)
	}
	// A subagent's entry names its card; the person's command is theirs.
	r = e.Search("TOKENIZER", 0)
	if len(r.Hits) != 1 || r.Hits[0].EntryID != "child" || r.Hits[0].ParentID != "agent" || r.Hits[0].Field != "text" {
		t.Fatalf("nested: %+v", r.Hits)
	}
	r = e.Search("scratch", 0)
	if len(r.Hits) != 1 || r.Hits[0].EntryID != "mine" || r.Hits[0].Sender == nil || r.Hits[0].Field != "detail" {
		t.Fatalf("person's command: %+v", r.Hits)
	}
	// A message's detail (a delivery error) is not searched; a step's is.
	if r = e.Search("No rule", 0); len(r.Hits) != 1 || r.Hits[0].Field != "detail" {
		t.Fatalf("step output: %+v", r.Hits)
	}
	// Blank queries match nothing; the limit caps and counts.
	if r = e.Search(" \n", 0); len(r.Hits) != 0 || r.More != 0 {
		t.Fatalf("blank: %+v", r)
	}
	if r = e.Search("build", 2); len(r.Hits) != 2 || r.More != 4 {
		t.Fatalf("limit: %d hits, %d more", len(r.Hits), r.More)
	}
	if r = e.Search("build", 10000); len(r.Hits) != 6 {
		t.Fatalf("over the maximum: %d", len(r.Hits))
	}
	// The route.
	h := &HTTP{Engine: e, Token: "private", Host: "127.0.0.1:18780", Origin: "http://127.0.0.1:18780", WebDir: t.TempDir()}
	req := httptest.NewRequest("GET", h.Origin+"/api/chats/search?q=build&limit=3", nil)
	req.Header.Set("Authorization", "Bearer private")
	w := httptest.NewRecorder()
	h.ServeHTTP(w, req)
	if w.Code != http.StatusOK {
		t.Fatalf("route: %d %s", w.Code, w.Body.String())
	}
	var res SearchResult
	if err := json.Unmarshal(w.Body.Bytes(), &res); err != nil {
		t.Fatal(err)
	}
	if len(res.Hits) != 3 || res.More != 3 || res.Hits[0].Field != "title" || res.Hits[2].EntryID != "n2" {
		t.Fatalf("route result: %+v", res)
	}
	req = httptest.NewRequest("GET", h.Origin+"/api/chats/search?q=", nil)
	req.Header.Set("Authorization", "Bearer private")
	w = httptest.NewRecorder()
	h.ServeHTTP(w, req)
	if w.Code != 200 || strings.TrimSpace(w.Body.String()) != `{"hits":[],"more":0}` {
		t.Fatalf("empty query: %d %s", w.Code, w.Body.String())
	}
	req = httptest.NewRequest("GET", h.Origin+"/api/chats/search?q=build", nil)
	w = httptest.NewRecorder()
	h.ServeHTTP(w, req)
	if w.Code != 401 {
		t.Fatalf("unauthenticated search: %d", w.Code)
	}
}

func TestSearchSnippets(t *testing.T) {
	text := []rune("The quick brown fox\njumps over the lazy dog and keeps on running far away")
	at := indexRunes(foldRunes(string(text)), []rune("lazy"))
	if got := snippetAt(text, at, at+4, 12); got != (Snippet{Before: "…over the ", Match: "lazy", After: " dog and…"}) {
		t.Fatalf("%+v", got)
	}
	if got := snippetAt([]rune("lazy dog"), 0, 4, 40); got != (Snippet{Match: "lazy", After: " dog"}) {
		t.Fatalf("%+v", got)
	}
	if got := snippetAt([]rune("a\n\nlazy"), 3, 7, 40); got.Before != "a " {
		t.Fatalf("%+v", got)
	}
	if indexRunes([]rune("abc"), []rune("abcd")) != -1 || indexRunes([]rune("abc"), nil) != -1 || indexRunes([]rune("xabc"), []rune("abc")) != 1 {
		t.Fatal("indexRunes")
	}
	if string(foldQuery("  Café\tAU ")) != "café au" {
		t.Fatalf("foldQuery: %q", string(foldQuery("  Café\tAU ")))
	}
}
